File: solutions/java/src/main/java/com/epi/ReverseLinkedListRecursive.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.
// @author Ivan Sharov

package com.epi;

class ReverseLinkedListRecursive {
  // @include
  public static ListNode<Integer> reverseLinkedList(ListNode<Integer> head) {
    if (head == null || head.next == null) {
      return head;
    }

    ListNode<Integer> newHead = reverseLinkedList(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
  }
  // @exclude

  private static <T> void print(ListNode<T> head) {
    if (head != null) {
      System.out.println("(" + head.data + ")");
      print(head.next);
    }
  }

  public static void main(String[] args) {
    ListNode<Integer> l1 = new ListNode<>(1, null);
    ListNode<Integer> l2 = new ListNode<>(2, null);
    l1.next = l2;
    ListNode<Integer> l3 = new ListNode<>(3, null);
    l2.next = l3;

    System.out.println("before reverse");
    print(l1);
    ListNode<Integer> newhead = reverseLinkedList(l1);
    System.out.println("\nafter reverse");
    print(newhead);
    newhead = reverseLinkedList(newhead);
    System.out.println("\nafter another reverse");
    print(newhead);
  }
}

File: static/python/reverse_linked_list_recursive.py
# Reverse_linked_list_recursive.cpp 69dab7578339012270872ba200cfff02f59ce894
from linked_list_prototype import ListNode


# @include
def reverse_linked_list(head):
    if not head or not head.next:
        return head
    new_head = reverse_linked_list(head.next)
    head.next.next = head
    head.next = None
    return new_head


# @exclude


def print_list(head):
    if head:
        print('(', head.data, ')', sep='')
        print_list(head.next)


def main():
    L1 = ListNode(1, ListNode(2, ListNode(3, None)))

    print('before reverse')
    print_list(L1)
    newhead = reverse_linked_list(L1)
    print('\nafter reverse')
    assert newhead.data == 3 and newhead.next.data == 2 and newhead.next.next.data == 1
    print_list(newhead)
    newhead = reverse_linked_list(newhead)
    assert newhead.data == 1 and newhead.next.data == 2 and newhead.next.next.data == 3
    print('\nafter another reverse')
    print_list(newhead)


if __name__ == '__main__':
    main()
